File: src/lib/services/tasks.ts
import { invoke } from "@tauri-apps/api/core";

export interface Task {
  id: number;
  name: string;
  total_seconds: number;
  created_at: string;
  note: string | null;
}

/**
 * Get today's date in YYYY-MM-DD format (using local timezone)
 */
export function getTodayDate(): string {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, "0");
  const day = String(now.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

/**
 * Get all tasks for a specific date (based on created_at timestamp)
 * @param date - Date in YYYY-MM-DD format
 */
export async function getTasksForDate(date: string): Promise<Task[]> {
  return invoke<Task[]>("get_tasks_for_date", { date });
}

/**
 * Get all tasks for today (convenience function)
 */
export async function getTodaysTasks(): Promise<Task[]> {
  return invoke<Task[]>("get_todays_tasks");
}

/**
 * Create a new task
 * @param name - Task name
 * @param date - Date in YYYY-MM-DD format. Task will be created with this date at midnight.
 * @param initialSeconds - Optional initial time in seconds (default: 0)
 */
export async function createTask(
  name: string,
  date: string,
  initialSeconds: number = 0
): Promise<Task> {
  return invoke<Task>("create_task", {
    name,
    date,
    initialSeconds: initialSeconds > 0 ? initialSeconds : null,
  });
}

/**
 * Update a task's total time
 */
export async function updateTaskTime(taskId: number, totalSeconds: number): Promise<void> {
  return invoke("update_task_time", { taskId, totalSeconds });
}

/**
 * Add seconds to a task's total time
 */
export async function addTimeToTask(taskId: number, secondsToAdd: number): Promise<void> {
  return invoke("add_time_to_task", { taskId, secondsToAdd });
}

/**
 * Adjust a task's time (can be positive or negative)
 */
export async function adjustTaskTime(taskId: number, secondsToAdjust: number): Promise<void> {
  return invoke("adjust_task_time", { taskId, secondsToAdjust });
}

/**
 * Get tasks for a date range (for history view, based on created_at)
 */
export async function getTasksInRange(startDate: string, endDate: string): Promise<Task[]> {
  return invoke<Task[]>("get_tasks_in_range", { startDate, endDate });
}

/**
 * Update a task's name
 */
export async function updateTaskName(taskId: number, newName: string): Promise<void> {
  return invoke("update_task_name", { taskId, newName });
}

/**
 * Delete a task
 */
export async function deleteTask(taskId: number): Promise<void> {
  return invoke("delete_task", { taskId });
}

/**
 * Get unique task names for autocomplete suggestions
 * Returns distinct task names ordered by most recently used
 */
export async function getUniqueTaskNames(): Promise<string[]> {
  return invoke<string[]>("get_unique_task_names");
}

/**
 * Get a single task by ID
 */
export async function getTaskById(taskId: number): Promise<Task | null> {
  return invoke<Task | null>("get_task_by_id", { taskId });
}

/**
 * Update a task's note
 */
export async function updateTaskNote(taskId: number, note: string | null): Promise<void> {
  return invoke("update_task_note", { taskId, note });
}

/**
 * Update a task's date
 * @param taskId - Task ID
 * @param newDate - Date in YYYY-MM-DD format
 */
export async function updateTaskDate(taskId: number, newDate: string): Promise<void> {
  return invoke("update_task_date", { taskId, newDate });
}

File: src-tauri/src/commands/tasks.rs
use crate::models::Task;
use crate::AppState;
use chrono::Local;
use tauri::State;

#[tauri::command]
pub fn get_tasks_for_date(state: State<AppState>, date: String) -> Result<Vec<Task>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::list_tasks_for_date(&conn, &date)
}

#[tauri::command]
pub fn get_todays_tasks(state: State<AppState>) -> Result<Vec<Task>, String> {
    let today = Local::now().format("%Y-%m-%d").to_string();
    get_tasks_for_date(state, today)
}

#[tauri::command]
pub fn create_task(
    state: State<AppState>,
    name: String,
    date: String,
    initial_seconds: Option<i64>,
) -> Result<Task, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::create_task(&conn, &name, &date, initial_seconds, None)
}

#[tauri::command]
pub fn update_task_time(
    state: State<AppState>,
    task_id: i64,
    total_seconds: i64,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::set_task_time(&conn, task_id, total_seconds)?;
    Ok(())
}

#[tauri::command]
pub fn add_time_to_task(
    state: State<AppState>,
    task_id: i64,
    seconds_to_add: i64,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::add_task_time(&conn, task_id, seconds_to_add)?;
    Ok(())
}

#[tauri::command]
pub fn adjust_task_time(
    state: State<AppState>,
    task_id: i64,
    seconds_to_adjust: i64,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::add_task_time(&conn, task_id, seconds_to_adjust)?;
    Ok(())
}

#[tauri::command]
pub fn get_tasks_in_range(
    state: State<AppState>,
    start_date: String,
    end_date: String,
) -> Result<Vec<Task>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::list_tasks_in_range(&conn, &start_date, &end_date)
}

#[tauri::command]
pub fn update_task_name(
    state: State<AppState>,
    task_id: i64,
    new_name: String,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::set_task_name(&conn, task_id, &new_name)?;
    Ok(())
}

#[tauri::command]
pub fn delete_task(state: State<AppState>, task_id: i64) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::delete_task(&conn, task_id)
}

#[tauri::command]
pub fn get_unique_task_names(state: State<AppState>) -> Result<Vec<String>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::unique_task_names(&conn)
}

#[tauri::command]
pub fn get_task_by_id(state: State<AppState>, task_id: i64) -> Result<Option<Task>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::get_task_by_id(&conn, task_id)
}

#[tauri::command]
pub fn update_task_note(
    state: State<AppState>,
    task_id: i64,
    note: Option<String>,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::set_task_note(&conn, task_id, note.as_deref())?;
    Ok(())
}

#[tauri::command]
pub fn update_task_date(
    state: State<AppState>,
    task_id: i64,
    new_date: String,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    time_tracker_core::set_task_date(&conn, task_id, &new_date)?;
    Ok(())
}
